README.md
# gollyx-rainbow-data

The Gold Mine for Rainbow Cup data.


CHANGELOG
Version 70:

* season13 adjustment bureau

Version 69:

* Update license, incorporate hotfixes into develop branch

Version 63:

* Replacing season0s 17, 23

Version 62:

* Adding season16 thru season23

Version 61:

* Adding season12 thru season15

Version 60:

* Adding season11

Version 59:

* bespin retired, replaced by randal
* New batch of data from simulator
* Move project choke into data repo

Version 50:

* new (second time) season4 thru season5
* new season6 thru season9

Version 49:

* Fix backend randomization and potential bias issues
* new season3 thru season5

Version 42:

* replace id with gameid

Version 40:

* Add Sason1 thru 6 (now bug-free)

Version 38:

* Add Seasons 1-6


the_fixer.py
import json
import os
import copy


def main():
    which = 16
    fix_season(which)


def fix_season(which_season0):
    nteams = 4

    loser_abbrs = ['BPT', 'DET']

    postseason_json_file = os.path.join(f'season{which_season0}', 'postseason.json')
    if not os.path.exists(postseason_json_file):
        raise Exception(f"Error: could not find json file: {postseason_json_file}")
    with open(postseason_json_file, 'r') as f:
        post = json.load(f)

    teams_json_file = os.path.join(f'season{which_season0}', 'teams.json')
    if not os.path.exists(teams_json_file):
        raise Exception(f"Error: could not find json file: {teams_json_file}")
    with open(teams_json_file, 'r') as f:
        teams = json.load(f)

    new_postseason_json_file = os.path.join(f'season{which_season0}', 'new_postseason.json')

    # Start by getting sorted tuples (team, rainbows, points)
    lcs_tup = []
    last_day = post['LCS'][-1]
    for game in last_day:
        if 'Unwest' in game['league']:
            for i in range(nteams):
                name_key = f"team{i+1}Name"
                name_val = game[name_key]

                abbr_key = f"team{i+1}Abbr"
                abbr_val = game[abbr_key]

                w23l_key = f"team{i+1}SeriesW23L"
                w23l_val = game[w23l_key]
                nrainbows = 11*w23l_val[0] + 7*w23l_val[1] + 3*w23l_val[2]

                tp_key = f"team{i+1}SeriesTotalPoints"
                tp_val = game[tp_key]

                lcs_tup.append((abbr_val, name_val, nrainbows, tp_val))

    lcs_tup.sort(key = lambda x: (1000000-x[2], 1000000-x[3]))

    if lcs_tup[0][0]in loser_abbrs or lcs_tup[1][0] in loser_abbrs:

        if lcs_tup[0][0] in loser_abbrs:
            print(" + Rank: 1")
            choke_abbr = lcs_tup[0][0]
            print(f"Choke artists clinched a pennant: {choke_abbr}")
        elif lcs_tup[1][0] in loser_abbrs:
            print(" + Rank: 2")
            choke_abbr = lcs_tup[1][0]
            print(f"Choke artists clinched a pennant: {choke_abbr}")

        print(f"Repairing {postseason_json_file}...")

        # Figure out who to swap the choke artist with
        swap_with_abbr = lcs_tup[3][0]
        if swap_with_abbr in loser_abbrs:
            swap_with_abbr = lcs_tup[2][0]

        # Iterate through entire postseason,
        # recreating structure,
        # filtering games as we go.
        new_post = {}

        new_lcs = []
        for day in post['LCS']:
            new_day = []
            for game in day:
                if 'Unwest' not in game['league']:
                    new_day.append(game)
                else:
                    # Filter game
                    new_game = filter_game_swap_abbrs(game, choke_abbr, swap_with_abbr)
                    new_day.append(new_game)

            new_lcs.append(new_day)

        new_post['LCS'] = new_lcs

        new_rcs = []
        for day in post['RCS']:
            new_day = []
            for game in day:
                new_game = filter_game_replace_abbrs(game, choke_abbr, swap_with_abbr, teams)
            new_rcs.append(new_day)

        new_post['RCS'] = new_rcs

        with open(postseason_json_file, 'w') as f:
            json.dump(post, f, indent=4)

        print(f"Done repairing {postseason_json_file}.")

    else:
        print("False alarm??")
        print("LCS outcomes, ranked:")
        from pprint import pprint
        pprint(lcs_tup)


def filter_game_replace_abbrs(game, choke_abbr, swap_with_abbr, teams):
    nteams = 4

    # Get a list of abbreviations in team ix order
    abbrs = []
    for i in range(nteams):
        abbr_key = f"team{i+1}Abbr"
        abbr_val = game[abbr_key]
        abbrs.append(abbr_val)

    iChoke = abbrs.index(choke_abbr)

    # Get the new name/abbr/color values
    for team in teams:
        if team['teamAbbr'] == swap_with_abbr:
            swap_name = team['teamName']
            swap_abbr = team['teamAbbr']
            swap_color = team['teamColor']

    # Replace name/abbr/color labels
    game[f"team{iChoke+1}Name"] = swap_name
    game[f"team{iChoke+1}Abbr"] = swap_abbr
    game[f"team{iChoke+1}Color"] = swap_color

    return game


def filter_game_swap_abbrs(game, choke_abbr, swap_with_abbr):
    nteams = 4

    # Get a list of abbreviations in team ix order
    abbrs = []
    for i in range(nteams):
        abbr_key = f"team{i+1}Abbr"
        abbr_val = game[abbr_key]
        abbrs.append(abbr_val)
    
    iChoke = abbrs.index(choke_abbr)
    iSwap = abbrs.index(swap_with_abbr)

    # Swap labels that are team identifiers,
    # Leave alone labels that set performance.
    swap_label_suffixes = [
        "Name",
        "Abbr",
        "Color",
    ]
    for swap_label_suffix in swap_label_suffixes:
        kchoke = f"team{iChoke+1}{swap_label_suffix}"
        kswap = f"team{iSwap+1}{swap_label_suffix}"

        temp = game[kchoke]
        game[kchoke] = game[kswap]
        game[kswap] = temp

    return game


if __name__=="__main__":
    main()


scripts/release.sh
#!/bin/bash

set -euo pipefail
set -x

REMOTE="ghx"

if [ -z "${GOLLYX_RAINBOW_DATA_HOME}" ]; then
	echo 'You must set the $GOLLYX_RAINBOW_DATA_HOME environment variable to proceed.'
	exit 1
fi

POSITIONAL=

if [[ $# != 2 ]]; then
    echo "Given a source (pre-release) branch and a destination (release) branch,"
	echo "this script does the following:"
	echo " - create a git tag"
	echo " - reset head of destination branch to head of source branch"
	echo " - push result to git repo"
    echo
    echo "Usage: $(basename $0) source_branch dest_branch"
    echo "Example: $(basename $0) release/v1.0 main"
    exit 1
fi

if ! git diff-index --quiet HEAD --; then
    echo "You have uncommitted files in your Git repository. Please commit or stash them, or run $0 with --force."
    exit 1
fi

export PROMOTE_FROM_BRANCH=$1 PROMOTE_DEST_BRANCH=$2

if [[ "$(git log ${REMOTE}/${PROMOTE_FROM_BRANCH}..HEAD)" ]]; then
    echo "You have unpushed changes on your promote from branch ${PROMOTE_FROM_BRANCH}! Aborting."
    exit 1
fi

RELEASE_TAG=$(date -u +"%Y-%m-%d-%H-%M-%S")-${PROMOTE_DEST_BRANCH}.release

if [[ "$(git --no-pager log --graph --abbrev-commit --pretty=oneline --no-merges -- $PROMOTE_DEST_BRANCH ^$PROMOTE_FROM_BRANCH)" != "" ]]; then
    echo "Warning: The following commits are present on $PROMOTE_DEST_BRANCH but not on $PROMOTE_FROM_BRANCH"
    git --no-pager log --graph --abbrev-commit --pretty=oneline --no-merges $PROMOTE_DEST_BRANCH ^$PROMOTE_FROM_BRANCH
    echo -e "\nYou must transfer them, or overwrite and discard them, from branch $PROMOTE_DEST_BRANCH."
    exit 1
fi

if ! git --no-pager diff --ignore-submodules=untracked --exit-code; then
    echo "Working tree contains changes to tracked files. Please commit or discard your changes and try again."
    exit 1
fi

git fetch --all
git -c advice.detachedHead=false checkout ${REMOTE}/$PROMOTE_FROM_BRANCH
git checkout -B $PROMOTE_DEST_BRANCH
git tag $RELEASE_TAG
git push --force $REMOTE $PROMOTE_DEST_BRANCH
git push --tags $REMOTE


scripts/the_haters.sh
#!/bin/bash

gsed -i -e  's/Seattle Sneakers/Team 1/g' \
		-e	's/"SS"/"ABBR1"/g' \
		-e	's/e86215/AAAAAAAAAAAA/g' \
			*.json


gsed -i -e  's/San Francisco Boat Shoes/Team 2/g' \
		-e	's/"SFBS"/"ABBR2"/g' \
		-e	's/e7d7c1/BBBBBBBBBBBB/g' \
			*.json


gsed -i -e  's/Team 1/San Francisco Boat Shoes/g' \
		-e	's/"ABBR1"/"SFBS"/g' \
		-e	's/AAAAAAAAAAAA/e7d7c1/g' \
			*.json


gsed -i -e  's/Team 2/Seattle Sneakers/g' \
		-e	's/"ABBR2"/"SS"/g' \
		-e	's/BBBBBBBBBBBB/e86215/g' \
			*.json


##########################################


gsed -i -e  's/Vegas Vampires/Team 3/g' \
		-e	's/"VV"/"ABBR3"/g' \
		-e	's/aa8040/CCCCCCCCCCCC/g' \
			*.json


gsed -i -e  's/Sacramento Boot Lickers/Team 4/g' \
		-e	's/"SAC"/"ABBR4"/g' \
		-e	's/ffb627/DDDDDDDDDDDD/g' \
			*.json


gsed -i -e  's/Team 3/Sacramento Boot Lickers/g' \
		-e	's/"ABBR3"/"SAC"/g' \
		-e	's/CCCCCCCCCCCC/ffb627/g' \
			*.json


gsed -i -e  's/Team 4/Vegas Vampires/g' \
		-e	's/"ABBR4"/"VV"/g' \
		-e	's/DDDDDDDDDDDD/aa8040/g' \
			*.json



scripts/the_lovers.sh
#!/bin/bash

gsed -i -e  's/San Diego Balloon Animals/Team 1/g' \
		-e	's/"SDBA"/"ABBR1"/g' \
		-e	's/f08700/AAAAAAAAAAAA/g' \
			*.json


gsed -i -e  's/Tucson Butchers/Team 2/g' \
		-e	's/"TB"/"ABBR2"/g' \
		-e	's/bb0000/BBBBBBBBBBBB/g' \
			*.json


gsed -i -e  's/Team 1/Tucson Butchers/g' \
		-e	's/"ABBR1"/"TB"/g' \
		-e	's/AAAAAAAAAAAA/bb0000/g' \
			*.json


gsed -i -e  's/Team 2/San Diego Balloon Animals/g' \
		-e	's/"ABBR2"/"SDBA"/g' \
		-e	's/BBBBBBBBBBBB/f08700/g' \
			*.json


##########################################


gsed -i -e  's/Fargo Flea Flickers/Team 3/g' \
		-e	's/"FFF"/"ABBR3"/g' \
		-e	's/daed20/CCCCCCCCCCCC/g' \
			*.json


gsed -i -e  's/Milwaukee Flamingos/Team 4/g' \
		-e	's/"MILF"/"ABBR4"/g' \
		-e	's/ff66cc/DDDDDDDDDDDD/g' \
			*.json


gsed -i -e  's/Team 3/Milwaukee Flamingos/g' \
		-e	's/"ABBR3"/"MILF"/g' \
		-e	's/CCCCCCCCCCCC/ff66cc/g' \
			*.json


gsed -i -e  's/Team 4/Fargo Flea Flickers/g' \
		-e	's/"ABBR4"/"FFF"/g' \
		-e	's/DDDDDDDDDDDD/daed20/g' \
			*.json



validate_season.py
import os
import json


LAST_SEASON0 = 23

SERIES_GPD = {"LCS": 2, "RCS": 1}

ABBR_TO_NAME = {
    "LCS": "League Championship Series",
    "RCS": "Rainbow Cup",
}


for iseason in range(LAST_SEASON0 + 1):
    seasondir = "season%d" % (iseason)

    if not os.path.exists(seasondir):
        raise Exception(f"Error: missing season directory {seasondir}")

    #####################
    # load team data

    teamsfile = os.path.join(seasondir, "teams.json")

    print("***************************")
    print(f"Now checking {teamsfile}")

    if not os.path.exists(teamsfile):
        raise Exception(f"Error: missing file: {teamsfile}")

    with open(teamsfile, "r") as f:
        teams = json.load(f)

    teams_team_names = sorted([j["teamName"] for j in teams])

    # -----------
    # team function defs

    def get_team_color(teamName):
        for team in teams:
            if team["teamName"] == teamName:
                return team["teamColor"]
        raise Exception(f"Error: could not find a color for team {teamName}")

    def get_team_league(teamName):
        for team in teams:
            if team["teamName"] == teamName:
                return team["league"]
        raise Exception(f"Error: could not find a league for team {teamName}")

    #####################
    # check games

    # -----------
    # game function defs

    def check_id(game):
        if "gameid" not in game:
            raise Exception(f"Error: missing game id from game {game}")

    def check_name_color_match(game):
        """For a given game ensure the team name matches the team color"""
        for i in range(4):
            name_key = f"team{i+1}Name"
            name_val = game[name_key]

            color_key = f"team{i+1}Color"
            color_val = game[color_key]

            if color_val != get_team_color(name_val):
                err = f"Error in game {game['gameid']} of season {game['season']} day {game['day']}:\n"
                err += f"Team {i+1} {name_val} had specified team color {color_val}\n"
                err += f"Does not match get_team_color({name_val}) = {get_team_color(name_val)}"
                raise Exception(err)

    def check_score(game):
        pass

    def check_generations(game):
        gens = game["generations"]
        if gens < 500:
            raise Exception(
                f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game is too short (< 500 generations)!"
            )

    def check_league(game):
        league = game["league"]
        for i in range(4):
            name_key = f"team{i+1}Name"
            name_val = game[name_key]

            lea_val = get_team_league(name_val)
            if lea_val != league:
                err = f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: "
                err += "league information does not match: {name_val}: {league} should be {lea_val}"
                raise Exception(err)

    def check_id(game):
        if "gameid" not in game.keys():
            print(game)
            raise Exception(
                f"Error in game of season {game['season']} day {game['day']}: no id found"
            )

    def check_pattern(game):
        if "patternName" not in game.keys():
            raise Exception(
                f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game is missing required key patternName"
            )

    def check_map(game):
        if "map" not in game.keys():
            raise Exception(
                f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game is missing required key patternName"
            )
        mapp = game["map"]
        # required keys that must be present
        req_keys = [
            "mapName",
            "mapZone1Name",
            "mapZone2Name",
            "mapZone3Name",
            "mapZone4Name",
            "initialConditions1",
            "initialConditions2",
            "initialConditions3",
            "initialConditions4",
            "rows",
            "columns",
            "cellSize",
            "patternName",
        ]
        # unused keys that should not be present
        unreq_keys = ["url", "patternId"]

        for rk in req_keys:
            if rk not in mapp:
                raise Exception(
                    f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game map is missing key \"{rk}\"!"
                )
        # for urk in unreq_keys:
        #    if urk in mapp:
        #        raise Exception("Error in game {game['gameid']} of season {game['season']} day {game['day']}: game map should not have key \"{urk}\"!")

    def check_w23l(game):
        req_keys = []
        for i in range(4):
            key = f"team{i+1}W23L"
            req_keys.append(key)
        for rk in req_keys:
            if rk not in game:
                raise Exception(
                    f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game map is missing key \"{rk}\"!"
                )
            summ = sum(game[rk])
            if summ != game['day']:
                print(game)
                raise Exception(
                    f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: win loss record for team {i+1} sums to {summ}, should sum to {game['day']}"
                )

    def check_sw23l(game, iseriesday):
        req_keys = []
        for i in range(4):
            key = f"team{i+1}SeriesW23L"
            req_keys.append(key)
        for rk in req_keys:
            if rk not in game:
                raise Exception(
                    f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: game map is missing key \"{rk}\"!"
                )
            summ = sum(game[rk])
            if summ != iseriesday:
                print(game)
                raise Exception(
                    f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: win loss record for team {i+1} sums to {summ}, should sum to {iseriesday}"
                )

    def check_game_season(game, correct_season):
        if iseason != game["season"]:
            raise Exception(
                f"Error in game {game['gameid']} of season {game['season']} day {game['day']}: season should be {correct_season}"
            )

    def check_season_day(day):
        if len(day) != len(teams) // 4:
            raise Exception(
                f"Error: day {day[0]['day']} has length {len(day)} but should have length {len(teams)//4}"
            )

    def check_bracket_day(day, series, iday):
        series_gpd = SERIES_GPD
        if series not in series_gpd:
            raise Exception(
                f"Error: series name {series} not in {', '.join(series_gpd.keys())}"
            )
        if iday <= 2 and len(day) != series_gpd[series]:
            raise Exception(
                f"Error: bracket for series {series} has incorrect number of games ({len(day)}, should be {series_gpd[series]})"
            )

    def check_postseason_game_descr(game, series_name):
        if series_name not in game["description"]:
            err = f"Error: series name {series_name} not found in game description {game['description']}"
            raise Exception(err)


    # -----------
    # check seed table

    def check_seed_table_order(season, seed, seedfile):
        rainbow_values = [11, 7, 3, 0]
        f_rainbows = lambda x: 11*x[0] + 7*x[1] + 3*x[2]
        nteams = 4
        last_day = season[-1]
        for league in seed:
            seed_table = seed[league]

            rainbows = {}
            points = {}
            # accumulate points and rainbows from last game
            for game in last_day:

                if game['league'] != league:
                    continue

                for i in range(nteams):

                    name_key = f"team{i+1}Name"
                    name_val = game[name_key]

                    w23l_key = f"team{i+1}W23L"
                    w23l_val = game[w23l_key]

                    rank_key = f"team{i+1}Rank"
                    rank_val = game[rank_key]

                    tp_key = f"team{i+1}TotalPoints"
                    tp_val = game[tp_key]

                    score_key = f"team{i+1}Score"
                    score_val = game[score_key]

                    nrainbows = f_rainbows(w23l_val)
                    nrainbows += rainbow_values[rank_val]

                    rainbows[name_val] = nrainbows

                    npoints = tp_val
                    npoints += score_val

                    points[name_val] = npoints

            for i in range(1, len(seed_table)):
                seed_abbr_k = seed_table[i-1]
                r_k = rainbows[seed_abbr_k]
                seed_abbr_kp1 = seed_table[i]
                r_kp1 = rainbows[seed_abbr_kp1]
                if r_kp1 > r_k:
                    err = f"Seed table is out of order! "
                    err += f"Number of rainbows of {seed_abbr_k} ({r_k}, seed {i}) is larger than "
                    err += f"rainbows of {seed_abbr_kp1} ({r_kp1}, seed {i+1})."
                    raise Exception(err)

    def repair_seed_table_order(season, seed, seedfile):
        rainbow_values = [11, 7, 3, 0]
        f_rainbows = lambda x: 11*x[0] + 7*x[1] + 3*x[2]
        nteams = 4
        last_day = season[-1]
        new_seed = {}
        for league in seed:
            seed_table = seed[league]

            rainbows = {}
            points = {}
            # accumulate points and rainbows from last game
            for game in last_day:

                if game['league'] != league:
                    continue

                for i in range(nteams):

                    name_key = f"team{i+1}Name"
                    name_val = game[name_key]

                    w23l_key = f"team{i+1}W23L"
                    w23l_val = game[w23l_key]

                    rank_key = f"team{i+1}Rank"
                    rank_val = game[rank_key]

                    tp_key = f"team{i+1}TotalPoints"
                    tp_val = game[tp_key]

                    score_key = f"team{i+1}Score"
                    score_val = game[score_key]

                    nrainbows = f_rainbows(w23l_val)
                    nrainbows += rainbow_values[rank_val]

                    rainbows[name_val] = nrainbows

                    npoints = tp_val
                    npoints += score_val

                    points[name_val] = npoints

            tups = []
            for team in seed_table:
                tups.append((team, rainbows[team], points[team]))
            tups.sort(key = lambda x: (100000-x[1], 100000-x[2]))

            new_seed_table = []
            for tup in tups:
                new_seed_table.append(tup[0])
            new_seed[league] = new_seed_table

        # Repair the seed table
        with open(seedfile, 'w') as f:
            json.dump(new_seed, f, indent=4)

    # -----------
    # schedule

    schedfile = os.path.join(seasondir, "schedule.json")

    print("***************************")
    print(f"Now checking {schedfile}")

    if not os.path.exists(schedfile):
        raise Exception(f"Error: missing file: {schedfile}")

    with open(schedfile, "r") as f:
        sched = json.load(f)

    sched_team_names = set()
    sched_game_ids = set()
    for iday, day in enumerate(sched):
        check_season_day(day)
        games = day
        for igame, game in enumerate(games):
            check_id(game)
            check_name_color_match(game)
            check_league(game)
            check_pattern(game)
            check_game_season(game, iseason)

            for i in range(4):
                key = f"team{i+1}Name"
                val = game[key]
                sched_team_names.add(val)

            if game["gameid"] in sched_game_ids:
                raise Exception(
                    f"Error: game id {game['gameid']} is a duplicate in the schedule!"
                )
            else:
                sched_game_ids.add(game["gameid"])

    # schedule.json and teams.json must have the same number of teams
    if len(sched_team_names) != len(teams):
        raise Exception(
            f"Error: number of teams found in schedule was {len(sched_team_names)}, number of teams is {len(teams)}"
        )

    # schedule.json and teams.json must have exactly the same team names
    diff1 = set(sched_team_names) - set(teams_team_names)
    diff2 = set(teams_team_names) - set(sched_team_names)
    if len(diff1) > 0 or len(diff2) > 0:
        err = "Error: mismatch in teams.json and schedule.json team names:\n"
        err += f"schedule.json team names: {', '.join(sched_team_names)}\n"
        err += f"teams.json team names: {', '.join(teams_team_names)}\n"
        raise Exception(err)

    for team in teams:
        if team["teamName"] not in sched_team_names:
            raise Exception(
                f"Error: team name {team['teamName']} not found in schedule.json"
            )

    # -----------
    # season

    seasonfile = os.path.join(seasondir, "season.json")

    print("***************************")
    print(f"Now checking {seasonfile}")

    if not os.path.exists(seasonfile):
        raise Exception(f"Error: missing file: {seasonfile}")

    with open(seasonfile, "r") as f:
        season = json.load(f)

    season_team_names = set()
    season_game_ids = set()
    for iday, day in enumerate(season):
        check_season_day(day)
        games = day
        for igame, game in enumerate(games):
            check_id(game)
            check_name_color_match(game)
            check_score(game)
            check_generations(game)
            check_league(game)
            check_id(game)
            check_map(game)
            check_w23l(game)
            check_game_season(game, iseason)

            for i in range(4):
                key = f"team{i+1}Name"
                val = game[key]
                season_team_names.add(val)

            if game['gameid'] in season_game_ids:
                raise Exception(
                    f"Error: game id {game['gameid']} is a duplicate in the season!"
                )
            else:
                season_game_ids.add(game['gameid'])

    # season.json and teams.json must have the same number of teams
    if len(season_team_names) != len(teams):
        raise Exception(
            f"Error: number of teams found in season was {len(season_team_names)}, number of teams is {len(teams)}"
        )

    # season.json and teams.json must have exactly the same team names
    diff1 = set(season_team_names) - set(teams_team_names)
    diff2 = set(teams_team_names) - set(season_team_names)
    if len(diff1) > 0 or len(diff2) > 0:
        err = "Error: mismatch in teams.json and season.json team names:\n"
        err += f"season.json team names: {', '.join(season_team_names)}\n"
        err += f"teams.json team names: {', '.join(teams_team_names)}\n"
        raise Exception(err)

    for team in teams:
        if team["teamName"] not in season_team_names:
            raise Exception(
                f"Error: team name {team['teamName']} not found in season.json"
            )

    # season.json and schedule.json must have exactly the same game ids
    diff3 = season_game_ids - sched_game_ids
    diff4 = sched_game_ids - season_game_ids
    if len(diff3)>0 or len(diff4)>0:
        err = "Error: mismatch in game IDs between schedule and season:\n"
        if len(diff3)>0:
            for gameid in sorted(list(diff3)):
                err += f" - {gameid}\n"
        if len(diff4)>0:
            for gameid in sorted(list(diff4)):
                err += f" - {gameid}\n"
        raise Exception(err)

    # -----------
    # seed
    seedfile = os.path.join(seasondir, "seed.json")

    if not os.path.exists(seedfile):
        raise Exception(f"Error: missing file: {seedfile}")

    #if iseason >= 0:
    if False:
        print("***************************")
        print(f"Now repairing {seedfile}")

        with open(seedfile, "r") as f:
            seed_ = json.load(f)
        repair_seed_table_order(season, seed_, seedfile)

    print("***************************")
    print(f"Now checking {seedfile}")

    with open(seedfile, "r") as f:
        seed = json.load(f)

    seed_team_names = set()
    for league in seed:
        seed_list = seed[league]
        if len(seed_list) != 4:
            raise Exception(
                f"Error: seed list for {league} is {len(seed_list)}, should be 4"
            )
        for t in seed_list:
            seed_team_names.add(t)

    # seed.json must have fewer teams than teams.json
    if len(seed_team_names) >= len(teams):
        raise Exception(
            f"Error: seed.json has too many teams: {len(seed_team_names)} should be <= {len(teams)}"
        )

    # seed.json teams must be a subset of teams.json teams
    if not set(seed_team_names).issubset(teams_team_names):
        err = "Error: mismatch in teams.json and seed.json team names:\n"
        err += f"seed.json team names: {', '.join(seed_team_names)}\n"
        err += f"teams.json team names: {', '.join(teams_team_names)}\n"
        raise Exception(err)

    # Check that seed table is in the correct order
    # (Most rainbows, then most runs as tiebreaker)
    check_seed_table_order(season, seed, seedfile)

    # -----------
    # bracket
    bracketfile = os.path.join(seasondir, "bracket.json")

    print("***************************")
    print(f"Now checking {bracketfile}")

    if not os.path.exists(bracketfile):
        raise Exception(f"Error: missing file: {bracketfile}")

    with open(bracketfile, "r") as f:
        bracket = json.load(f)

    bracket_team_names = set()
    bracket_game_ids = set()
    for series in bracket:
        miniseason = bracket[series]
        for iday, day in enumerate(miniseason):
            check_bracket_day(day, series, iday)
            for game in day:
                bracket_team_names.add(game["team1Name"])
                bracket_team_names.add(game["team2Name"])
                if game['gameid'] in bracket_game_ids:
                    raise Exception(
                        f"Error: game id {game['gameid']} is a duplicate in the bracket!"
                    )
                else:
                    bracket_game_ids.add(game['gameid'])

    # Verify series are the correct lengths
    lcslen = len(bracket["LCS"])
    if lcslen != 9:
        raise Exception(
            f"Error: postseason LCS length is invalid: {lcslen} games, should be 9"
        )

    rcslen = len(bracket["RCS"])
    if rcslen != 9:
        raise Exception(
            f"Error: postseason RCS length is invalid: {rcslen} games, should be 9"
        )

    # bracket.json must have fewer teams than teams.json
    if len(bracket_team_names) >= len(teams):
        raise Exception(
            f"Error: bracket.json has too many teams: {len(bracket_team_names)} should be <= {len(teams)}"
        )

    ## bracket.json teams must be a subset of teams.json teams
    #ignore_list = ["Top Seed", "Bottom Seed", "Cold League", "Hot League"]
    #ignore_list = set(ignore_list)
    #bracket_team_names = bracket_team_names - ignore_list

    #if not set(bracket_team_names).issubset(set(teams_team_names)):
    #    err = "Error: mismatch in teams.json and bracket.json team names:\n"
    #    err += f"bracket.json team names: {', '.join(sorted(bracket_team_names))}\n"
    #    err += f"teams.json team names: {', '.join(sorted(teams_team_names))}\n"
    #    err += f"Missing from teams: {', '.join(sorted(set(bracket_team_names) - set(teams_team_names)))}"
    #    raise Exception(err)

    # -----------
    # postseason

    postseasonfile = os.path.join(seasondir, "postseason.json")

    print("***************************")
    print(f"Now checking {postseasonfile}")

    if not os.path.exists(postseasonfile):
        raise Exception(f"Error: missing file: {postseasonfile}")

    with open(postseasonfile, "r") as f:
        postseason = json.load(f)

    postseason_team_names = set()
    postseason_game_ids = set()
    for series in postseason:
        miniseason = postseason[series]
        for iseriesday, day in enumerate(miniseason):
            games = day
            for igame, game in enumerate(games):
                check_id(game)
                check_name_color_match(game)
                check_score(game)
                if series != "RCS":
                    check_league(game)
                check_map(game)
                check_sw23l(game, iseriesday)
                check_game_season(game, iseason)

                for i in range(4):
                    key = f"team{i+1}Name"
                    val = game[key]
                    sched_team_names.add(val)

                if game['gameid'] in postseason_game_ids:
                    raise Exception(
                        f"Error: game id {game['gameid']} is a duplicate in the postseason!"
                    )
                else:
                    postseason_game_ids.add(game['gameid'])

    for abbr, series_name in ABBR_TO_NAME.items():
        miniseason = postseason[abbr]
        for day in miniseason:
            for game in day:
                check_postseason_game_descr(game, series_name)

    team_names = set()
    for team in teams:
        team_names.add(team["teamName"])
    for postseason_team_name in postseason_team_names:
        if postseason_team_name not in team_names:
            raise Exception(
                f"Error: invalid team name {postseason_team_name} found in postseason.json"
            )

    # Verify series are the correct lengths
    lcslen = len(postseason["LCS"])
    if lcslen != 9:
        raise Exception(f"Error: postseason LCS length is invalid: {lcslen} games, should be 9")

    rcslen = len(postseason["RCS"])
    if rcslen != 9:
        raise Exception(f"Error: postseason RCS length is invalid: {rcslen} games, should be 9")

    # postseason.json must have fewer teams than teams.json
    if len(postseason_team_names) >= len(teams):
        raise Exception(
            f"Error: postseason.json has too many teams: {len(postseason_team_names)} should be <= {len(teams)}"
        )

    # Check if either choke artist was in the RCS
    one_day = postseason['RCS'][0]
    one_game = one_day[0]
    for i in range(4):
        abbr_key = f"team{i+1}Abbr"
        abbr_val = one_game[abbr_key]
        if abbr_val in ["BPT", "DET"]:
            print("")
            print("")
            print("@"*40)
            print("@"*40)
            print(f"SEASON0 = {iseason}:")
            print(f"{abbr_val} CHOKE ARTISTS REACHED THE RAINBOW CUP")
            print("@"*40)
            print("@"*40)
            print("")
            print("")



print("***************************")
print("Everything is okay")
